### ISSUE TITLE
Mitigations per "Self-encrypting deception: weaknesses in the encryption of solid state drives (SSDs)"

### ISSUE THREAD

**Issue by vogelfreiheit** [Drive-Trust-Alliance/sedutil#260]
Might want to take a look at this:

https://www.ru.nl/publish/pages/909282/draft-paper.pdf

Most relevant bits are those related to the models that populate EMPTY passwords for the master record:

> As stated in Section II-B, the MASTER PASSWORD CAPABILITY
> bit determines whether the factory-set Master password
> may unlock the drive. In order for the end user to redeem
> himself from the Master password acting as a security bypass
> mechanism, he either has to set the MASTER PASSWORD
> CAPABILITY bit to Maximum, or change the Master password.
> In the case of the MX300, the former approach is insufficient.
> We found that the Master password allows for successful
> decryption of the RDS key, regardless of the MASTER PASSWORD
> CAPABILITY bit. Hence, in case the end user has set it
> to Maximum, but has not changed the Master password, the
> drive’s contents is still accessible to anyone in possession of
> the default Master password. In the case of the MX300, this
> is an empty string.

As a potential workaround, sedutil should warn users (at very least) to change default Master passwords, as well as motivate them to check that in fact any password present is intentional.

Honestly, SED was expected to be intentionally broken from the get go, and the paper might not have covered all of the dirt to be found in the firmware for the different models.  Any efforts to at least make obviously unwise configurations more obvious are worth doing, IMO.

Gotta get some lawyers with sharp teeth involved in a class action or two against shady vendors.

**Comment by boldandbusted** [comment]
Howdy. Sorry if I'm showing how dense I am, but I've got some questions regarding this. :) I set up my drive with the old 'msed' (v0.23beta). I set a decryption (SID) password that I enter with every boot. However, in light of this paper and it's assertions, is that not enough to protect my data against decryption? Is there a non-dangerous way to determine if my drive is vulnerable, from within the Linux booted from the encrypted drive? Also, can I use the current code against a drive configured with older (and now deprecated) software? Thanks in advance.

**Comment by brendanhoar** [comment]
The biggest issues reported in the paper for the more contemporary OPAL drives tested were with the ATA Security (aka Class 0) configurations. 

My takeaway is:
1. They only tested Crucial(Micron) and Samsung drives, so not much can be gleaned on what other manufacturers have implemented.
2. Don't use Crucial drives, they appeared to be a disaster area.
3. Use only Samsung drives, model 840 or later models.
4. If using ATA Password, CHECK THE SECURITY LEVEL USING HDPARM and look for the maximum flag being set.  The use of user vs master passwords and the use of high vs. maximum flag are not correlated (but have impacts on each other). The result may be mainboard BIOS-dependent. Side note: I misunderstood this after first reading the paper and started setting both the user and master password. On a Lenovo Thinkpad W520, that *decreased* the security because when both are set, the BIOS sets the flag to high security. If, however, you set only the User password, the BIOS sets the maximum flag. The maximum flag ensures the key is always bound to the password and encrypted with it. With high, that is not the case, at least with the Samsung drives.
5. If using OPAL with Samsung 840 or later (such as via sedutil) the authors did not find any alarming security vulnerabilities. I find that somewhat comforting.

**Comment by dtasupport** [comment]
I responded to this paper at http://www.privust.com/sedlies/  .
www.privust.com is a link to a book on Amazon I wrote this year that
describes the history and security / privacy model behind SEDs.
regards,

Bob Thibadeau
412 370 1245

**Comment by brendanhoar** [comment]
Thanks for your thoughts, Bob. They are illuminating.

Separating the keys from the memory space of the main general purpose computer 

Since OPAL SEDs hit the market, I have always recommended the two-layer approach to disk encryption: open-source OS-level software encryption, many eyes sometimes works, but this vulnerable to OS-level software flaws; standards-based hardware encryption, keys always separate from OS and PBA erasing PW from memory before chaining boot, but this is vulnerable to poor implementation by manufacturers. Open-source purists can rant all they want about untrustable manufacturers (cf. a certain party on Qubes mailing list who replies to my every post). But THAT IS A KNOWN ISSUE. Plan for that. Defense in depth. Researchers have your back too, if perhaps a little later than we'd like.

Also, caveat on the Samsung 840/850 ATA Password flaws: after reviewing manufacturer documentation as well as discussion on forums such as hddguru and hddoracle, one may be able to argue that the behavior (lack of key encryption) of ATA Password security level = high on SED drives is as intended.

However, BIOS on-screen documentation should be explicit about the security defaults used and risks. BIOS should also give the end user the capability to choose high vs. maximum instead of silently choosing for the user.  Consumer documentation that comes from drive manufacturers should also cover this. None of that will happen, of course.

So, thanks to you (and Mike) for OPAL as the way forward.

I am curious if Samsung's most recent firmware releases have modified their internal ATA Password methodology to fully encrypt the key even with security set to high, just to give the end-user the benefit of the doubt. I think automatically updating the key slots after the firmware upgrade, if a user password is set, would be feasible - extract the master password hash, encrypt the plaintext key with it, validate it, erase the plaintext key, erase the master password hash. When testing external-supplied passwords to decrypt the key, hashing the entered passwords would be required...but they may do that already for both passwords anyway.

B

**Comment by dtasupport** [comment]
Yes.  Unfortunately, the work of the Trusted Computing Group stalled
precisely when they should have gotten busy on standardizing how the drives
cryptographically handled the passcodes...and the spec allows public key
(the core spec), but that was not made mandatory for Opal (which it should
be).  You may be interested to look for the Protection Profile for 'full
disk encryption' published a few years ago which shows a technique I came
up with for confirming that a drive is indeed encrypting properly (i.e.,
and not lying about it).  The biggest problem I have encountered is that
engineers and marketing people often don't know anything about computer
security and think they do, and think they can play with it, when they
can't...this is still happening today...it requires customers to scream at
them to get it right.  Seagate, when I was there anyway, was very careful.
Samsung was careful, I think.  Certifications like FIPS and Common Criteria
mean virtually nothing if the manufacturers are not respectful of getting
the security right and being transparent about what they are doing.
regards,

Bob Thibadeau
412 370 1245
